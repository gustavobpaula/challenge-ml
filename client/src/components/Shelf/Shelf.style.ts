import styled from '@emotion/styled'

export const ShelfList = styled.ul`
  list-style: none;
  padding: ${({ theme }) => theme.spacings.rule1};
`

export const ShelfListItem = styled.li`
  border-bottom: ${({ theme }) => `2px solid ${theme.colors.gray4}`};

  &:last-of-type {
    border-bottom: none;
  }
`
