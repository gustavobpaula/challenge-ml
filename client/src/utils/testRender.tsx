import React from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'

import theme from 'config/theme'

type RenderComponentProps = {
  component: React.ReactNode
  useTheme?: boolean
  options?: RenderOptions
}

type ProviderProps = {
  children: React.ReactNode
}

const Theme = ({ children }: ProviderProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const RenderComponent = ({
  component,
  useTheme = false,
  options = {},
}: RenderComponentProps) => {
  const providers = [useTheme && Theme]

  const componentWithProviders = providers.reduce((accComponent, Provider) => {
    if (!Provider) return accComponent

    return <Provider>{accComponent}</Provider>
  }, component)

  return render(<>{componentWithProviders}</>, options)
}
