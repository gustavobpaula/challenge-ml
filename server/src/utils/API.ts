import fetch from 'node-fetch'
import { URLSearchParams } from 'url'

const ML_API = 'https://api.mercadolibre.com'

const API = {
  /**
   * Return currency object
   */
  getCurrency: async (id: string) => {
    try {
      const response = await fetch(`${ML_API}/currencies/${id}`)
      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
  /**
   * Return items by query
   */
  search: async (query: string) => {
    const params = new URLSearchParams()
    params.append('q', query)

    try {
      const response = await fetch(
        `${ML_API}/sites/MLA/search?${params.toString()}`
      )

      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
}

export default API
