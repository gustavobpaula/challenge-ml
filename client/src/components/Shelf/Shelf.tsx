import { memo } from 'react'

import { ShelfItem } from 'components'
import { ItemProps } from 'typings/APIProps'

import * as S from './Shelf.style'

type ShelfProps = {
  items?: ItemProps[]
}

function Shelf({ items }: ShelfProps) {
  return (
    <S.ShelfList aria-label="Shelf">
      {items?.map(item => {
        return (
          <S.ShelfListItem key={item.id}>
            <ShelfItem item={item} />
          </S.ShelfListItem>
        )
      })}
    </S.ShelfList>
  )
}

export default memo(Shelf)
