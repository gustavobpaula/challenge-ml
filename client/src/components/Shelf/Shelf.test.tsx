import { screen } from '@testing-library/react'

import { RenderComponent } from 'utils'
import Shelf from './Shelf'

const makeSut = () => {
  return RenderComponent({
    component: <Shelf />,
    useTheme: true,
  })
}

describe('Shelf component', () => {
  it('Should render match snapshot', () => {
    const { asFragment } = makeSut()

    expect(asFragment()).toMatchSnapshot()
  })

  it('Should render a shelf list', () => {
    makeSut()

    expect(screen.getByRole('list', { name: /shelf/i })).toBeInTheDocument()
  })
})
