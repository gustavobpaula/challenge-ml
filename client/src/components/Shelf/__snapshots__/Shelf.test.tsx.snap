// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shelf component Should render match snapshot 1`] = `
<DocumentFragment>
  <ul
    aria-label="Shelf"
    class="css-1f5equ3"
  />
</DocumentFragment>
`;
